preserve node meta to fix inductor generated kernel name for pattern matched graphs

jazlyn5 opened the pull request:
Summary:
When using inductor pattern matcher to replace graphs, the graph generated by replacement function can be missing `original_aten` metadata for the replaced nodes.  This further results in inductor failing to generate a sensible kernel name, eg. `tri_poi_fused_0` , missing the aten op name.

This diff attempts to fix that by allowing tracing the graph in replacement function with `preserve_node_meta`. Included this as an option to turn on in `pattern_matcher.fwd_only` function.

Can confirm that with the fix, MTIA's pattern matcher replaced original graph with a node that has original_aten meta, and inductor generated kernel name has op name.

Test Plan:
added kernel_name check to afg_inductor_test silu test

Rollback Plan:

Differential Revision: D80183670




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -2100,15 +2101,17 @@ def fwd_only(
-            get_decomp_fn() if get_decomp_fn is not None else select_decomp_table()
-        )
-        gm = make_fx(fn, decompositions, tracing_mode="real")(*args)
+        with preserve_node_meta() if should_preserve_node_meta else contextlib.nullcontext():
```
cc @zou3519 any reason we wouldn't want this on always ?

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -2100,15 +2101,17 @@ def fwd_only(
     args: Sequence[Any],
     *,
     run_functional_passes: bool = True,
+    should_preserve_node_meta: bool = False,
```
Looks reasonable - cc @bdhirsh, @zou3519 any thoughts on if this should be on more generally ?

jazlyn5 committed:
preserve node meta to fix inductor generated kernel name for pattern matched graphs (#160542)

Summary:

When using inductor pattern matcher to replace graphs, the graph generated by replacement function can be missing `original_aten` metadata for the replaced nodes.  This further results in inductor failing to generate a sensible kernel name, eg. `tri_poi_fused_0` , missing the aten op name.

This diff attempts to fix that by allowing tracing the graph in replacement function with `preserve_node_meta`. Included this as an option to turn on in `pattern_matcher.fwd_only` function.

Can confirm that with the fix, custom pattern matcher replaced original graph with a node that has original_aten meta, and inductor generated kernel name has op name.

Test Plan:
added kernel_name check to afg_inductor_test silu test

oss test:
```
caffe2/test/inductor:pattern_matcher -- test_fwd_only_generate_original_aten_meta
```

Rollback Plan:

Differential Revision: D80183670

jazlyn5 commented:
@pytorchbot merge
